packages/core/src/parameters/getParameterDefinitionsFromSource.js

```diff
@@ -0,0 +1,174 @@
+const JSON5 = require('json5')
+
+const countSpaces = (l) => {
+  let count = 0
+  for (let i = 0; i < l.length; i++) {
+    if (l[i] === ' ') count++
+    else if (l[i] === '\t') count += 2
+    else break
+  }
+  return count
+}
+
+const getParameterDefinitionsFromSource = (script) => {
+  const lines = []
+  script.split('\n').forEach((l, i) => {
+    const trim = l.trim()
+    if (trim) {
+      lines.push({ code: trim, line: l, lineNum: i + 1, indent: countSpaces(l) })
+    }
+  })
+
+  let i = 0; let lineNum; let code; let prev; let prevIndent
+  while (i < lines.length) {
+    code = lines[i].code
+    i++
+    if (code.length > 12 && code.indexOf('@jscad-params') !== -1) break
+  }
+
+  let groupIndex = 1
+  const defs = []
+
+  while (i < lines.length) {
+    code = lines[i].code
+    lineNum = lines[i].lineNum
+    if (code[0] === '}') break
```

Comment:
So... if I understand correctly, this is the condition that says 'END OF PARAM DEFINITIONS'. Doesn't this assuming a certain coding style?

Reply:
yes, it expects closing bracket to be in separate line. and this whole thing is definitely for when there is one parameter per line.

```js
function main({//@jscad-params
  radius = 21
}){
  return sphere({radius})
}
// or
let defaults = {
//@jscad-params
  radius = 21
}

```

this would break current parser.

```js
function main({//@jscad-params
  radius = 21}){
  return sphere({radius})
}
```
I will add a check if the closing bracket is not in separate line, but in the same line as last parameter,

This whole concept must rely on certain style of defining the parameters, and `getParameterDefinitions` should be used when not wanting this style.

Reply:
> yes, it expects closing bracket to be in separate line. and this whole thing is definitely for when there is one parameter per line.

ok. just checking. please keep this information for the documentation.

Reply:
added check to handle closing bracket in same line with last  parameter. It is not an exhaustive check, but will help in most cases.